ca/certificate-authority.go

```diff
@@ -172,8 +190,47 @@ func loadKey(keyConfig cmd.KeyConfig) (priv crypto.Signer, err error) {
 	return
 }
 
+// checkHSMFault checks whether there has been an HSM fault observed within the
+// timeout window.  CA methods that use the HSM should call this method right
+// away, to minimize the performance impact of HSM outages.
+func (ca *CertificateAuthorityImpl) checkHSMFault() error {
+	ca.hsmFaultLock.Lock()
+	defer ca.hsmFaultLock.Unlock()
+
+	// If no timeout is set, never gate on a fault
+	if ca.hsmFaultTimeout == 0 {
+		return nil
+	}
+
+	now := ca.Clk.Now()
+	timeout := ca.hsmFaultLastObserved.Add(ca.hsmFaultTimeout)
+	if now.Before(timeout) {
+		return core.ServiceUnavailableError("HSM is unavailable")
+	}
+	return nil
+}
+
+// noteHSMFault updates the CA's state with regard to HSM faults.  CA methods
+// that use an HSM should pass errors that might be HSM errors to this method.
+func (ca *CertificateAuthorityImpl) noteHSMFault(err error) {
+	ca.hsmFaultLock.Lock()
+	defer ca.hsmFaultLock.Unlock()
+
+	if err != nil {
+		ca.stats.Inc(metricHSMFaultObserved, 1, 1.0)
+		ca.hsmFaultLastObserved = ca.Clk.Now()
+	}
+	return
+}
+
 // GenerateOCSP produces a new OCSP response and returns it
 func (ca *CertificateAuthorityImpl) GenerateOCSP(xferObj core.OCSPSigningRequest) ([]byte, error) {
+	if err := ca.checkHSMFault(); err != nil {
+		ca.log.WarningErr(err)
```

Comment:
You should now be able to move this log statement and the metric increment below it into `checkHSMFault`. Same below.